doc/sphinx/Pacemaker_Development/components.rst

```diff
@@ -5,6 +5,80 @@ The Pacemaker code can be intricate and difficult to follow. This chapter has
 some high-level descriptions of how individual components work.
 
 
+.. index::
+   single: controller
+   single: pacemaker-controld
+
+Controller
+##########
+
+``pacemaker-controld`` is the Pacemaker daemon that utilizes the other daemons
+to orchestrate actions that need to be taken in the cluster. It receives CIB
+change notifications from the CIB manager, passes the new CIB to the scheduler
+to determine whether anything needs to be done, uses the executor and fencer to
+execute any actions required, and sets failure counts (among other things) via
+the attribute manager.
+
+As might be expected, it has the most code of any of the daemons.
+
+.. index::
+   single: join
+
+Join sequence
+_____________
+
+Most daemons track their cluster peers using Corosync's membership and CPG
+only. The controller additionally makes peers run a gauntlet referred to as
+"joining" (also the join sequence or join process). The join sequence allows
+the DC to ensure nodes are ready to be assigned tasks.
+
+A node's current join phase is tracked by the ``join`` member of ``crm_node_t``
+(used in the peer cache). It is an ``enum crm_join_phase`` that (ideally)
+progresses from the DC's point of view as follows:
+
+* The node initially starts at ``crm_join_none``
+* Once the DC sends the node a join offer, it proceeds to
+  ``crm_join_welcomed``.
+
+  This can happen in two ways: the DC's peer status callback notices when the
+  node joins the messaging layer (which registers ``I_NODE_JOIN`` ->
+  ``A_DC_JOIN_OFFER_ONE`` -> ``do_dc_join_offer_one()`` ->
+  ``join_make_offer()`` to unconfirmed nodes); and/or the joining node sends a
+  join announce at its controller startup.
```

Comment:
I don't see why we do both, as opposed to the client simply requesting to join. (The DC making an offer wouldn't be sufficient, because for example a node could restart Pacemaker without leaving the messaging layer membership.) I suspect there is a good reason though.

Side note: I was also surprised to find while working on this code that the following scenario can happen:
1. DC makes offer
2. Client accepts offer
3. DC nacks client

Intuitively, you'd think that if the DC wouldn't make an offer if it were going to revoke it immediately... feels like a "high five, down low, too slow." Now that I know the DC makes the offer when it detects a join at the messaging layer, I realize it must have no idea whether the node is truly eligible for membership. I guess it's just the connotation of "offer" that bugged me.